### DISCUSSION TITLE
Performance questions

### DISCUSSION THREAD

**Discussion by caelandb** [fyne-io/fyne#4292] [Q&A] [answered]
Hi, I would like to start by saying I am trying to improve my programming knowledge.
I am working on a program that is in essence a custom terminal window (not related to the OS at all) that allows custom commands to be defined and issued (all written in Go).

However, I am running into issues with displaying output (with text styling, and many lines of output).
I have tried a number of approaches, but all have ended with really bad performance.
Any of the approaches that I have tried within a Scroll container are almost unusable.
I have decided instead to implement scrolling with KeyboardShortcuts and a TextGrid which is usable but still immensely taxing on the CPU.
After using pprof, I see that a lot of the work is coming from repaintWindow, WalkTrees, drawText and similar functions

What should I be looking into to try improve performance? I am keen to learn, but don't really know where to go from here.

Any help is appreciated

(Here is a picture of what the output display looks like)
![Untitled](https://github.com/fyne-io/fyne/assets/144162497/ee4f5c59-3217-4358-a6fb-56980b0316ac)

**Comment by andydotxyz** [comment, marked as answer]
Always refresh the smallest possible scope. If a single character changed then set/refresh only that cell.
Most performance issues in GUI apps come where developers just refresh the whole window, or large widgets, instead of the required scope.

Also consider checking out the https://GitHub.com/fyne-io/terminal project where there is a full terminal app and also re-usable Fyne widget that may save you a lot of time.

**Reply by caelandb** [reply]
Thank you!
